pilot/pkg/config/kube/ingress/conversion.go

```diff
@@ -189,6 +190,17 @@ func ConvertIngressVirtualService(ingress v1beta1.Ingress, domainSuffix string,
 		if f {
 			vs := old.Spec.(*networking.VirtualService)
 			vs.Http = append(vs.Http, httpRoutes...)
+			sort.SliceStable(vs.Http, func(i, j int) bool {
+				r1 := vs.Http[i].Match[0].Uri
+				r2 := vs.Http[j].Match[0].Uri
+				_, r1Ex := r1.MatchType.(*networking.StringMatch_Exact)
+				_, r2Ex := r2.MatchType.(*networking.StringMatch_Exact)
+				// TODO: default at the end
```

Comment:
Should we use the same logic as in CombineVHostRoutes? Otherwise I think we are going to get differing behavior for Ingress and normal Istio configs. Consider if I have a VS like
```
match:
- prefix: /
- exact: /foo
```

In sidecar and Gateway this will always match the prefix, but for Ingress it will match foo, which is changing the behavior?

Reply:
Probably - but it's a separate (and larger) issue, and maybe not safe for backporting. 
 
Ingress as it is now doesn't work without this fix. The problem is Ingress objects are merged
in the ingress controller - before CombineVHostRoutes.
With this change, the 'synthetic' ingress appears as a VirtualService with the exact rules first,
even if they are 'contributed' by different Ingress objects.

Reply:
I think best option is to wait for the new K8S Gateway - and what semantics it defines - and 
adjust ours to match.

Reply:
Can we just insert the matches from the autogenerated VS into the new spots instead of sorting the entire thing? If we sort the user defined VirtualService then we change the behavior. So instead of turning the example above into
```
match:
- exact: /foo
- exact: /some-auto-generated-route
- prefix: /
```

it becomes
```
match:
- exact: /some-auto-generated-route
- prefix: /
- exact: /foo
```

Reply:
I think it's the same thing - is a stable sort so the 'prefix' matches stay in the original order ( which is by creation time really ).

Reply:
But you are also modifying the user VirtualService, not just the autogenerated one

Reply:
The golden file changing is exactly my point

Reply:
How ? The golden file is for merged K8S Ingress - that was the point of making it in the ingress converter.

Reply:
The test is for converting K8S Ingress to VirtualService ( in istioctl )

Reply:
If I define a VirtualService with config:

```
match:
- prefix: /prefix
- exact: /foo
```

Then the merged VS should keep those in the same order. It can insert things at any point (like the acme stuff) but it shouldn't make the `exact: /foo` before `prefix: /prefix`.

I am a bit concerned that it would be changing behavior, but even more concerned that it changes behavior ONLY for Ingess, not Gateway or sidecars

Reply:
If you define a VirtualService - the ingress controller will not touch it. It only handles Ingress objects. That's why the change is in the ingress control - to make sure it is not changing any behavior for anything except Ingress objects.

How to merge Istio VirtualServices is an important topic - but I am not touching it in this PR. 
The Ingress controller creates a VirtualService from multiple Ingress objects - and AFAIK there 
is no defined merging semantic.

However many servers (and servlets) are matching exact paths first - and when you have 
multiple Ingress objects there is no defined order. Which means current behavior is simply 
broken - a /foo/* prefix before /foo/bar is clearly not intended behavior. 

Probably we should also reverse sort the exact matches - so /foo/bar/* is before /foo/* - but 
this is less common, and could be handled in a separate PR.

Reply:
sorry I think I misunderstood this as merging user vs with ingress, but it's just merging just generated ones so should be fine